src/Agent/NewRelic/Agent/Core/Agent.cs

```diff
@@ -406,33 +406,46 @@ public void RecordSupportabilityMetric(string metricName, int count)
             _agentHealthReporter.ReportSupportabilityCountMetric(metricName, count);
         }
 
-        public void RecordLogMessage(string frameworkName, DateTime timestamp, string logLevel, string logMessage, string spanId, string traceId)
+        public void RecordLogMessage(string frameworkName, object logEvent, Func<object,DateTime> getTimestamp, Func<object,object> getLogLevel, Func<object,string> getLogMessage, string spanId, string traceId)
         {
             _agentHealthReporter.ReportLogForwardingFramework(frameworkName);
 
-            var normalizedLevel = string.IsNullOrWhiteSpace(logLevel) ? "UNKNOWN" : logLevel.ToUpper();
+            var normalizedLevel = string.Empty;
+            if (_configurationService.Configuration.LogMetricsCollectorEnabled ||
+                _configurationService.Configuration.LogEventCollectorEnabled)
+            {
+                var logLevel = getLogLevel(logEvent).ToString();
+                normalizedLevel = string.IsNullOrWhiteSpace(logLevel) ? "UNKNOWN" : logLevel.ToUpper();
+            }
+
             if (_configurationService.Configuration.LogMetricsCollectorEnabled)
             {
                 _agentHealthReporter.IncrementLogLinesCount(normalizedLevel);
             }
 
             // IOC container defaults to singleton so this will access the same aggregator
-            if (!_configurationService.Configuration.LogEventCollectorEnabled || string.IsNullOrWhiteSpace(logMessage))
+            if (_configurationService.Configuration.LogEventCollectorEnabled) 
             {
-                return;
-            }
+                var logMessage = getLogMessage(logEvent);
+                if (string.IsNullOrWhiteSpace(logMessage))
+                {
+                    return;
+                }
+                var timestamp = getTimestamp(logEvent).ToUniversalTime();
```

Comment:
Couldn't you ToUnixTimeMilliseconds() the time here and just store it as a value type?

Reply:
Good idea!